File: 004-Bloom-Filter/bloom_filter.py
import mmh3
from bitarray import bitarray
import json

class BloomFilter:
    def __init__(self, size, hash_functions):
        self.size = size
        self.hash_functions = hash_functions
        self.bit_array = bitarray(size)
        self.bit_array.setall(0)

    def add(self, item):
        for seed in range(self.hash_functions):
            index = mmh3.hash(item, seed) % self.size
            self.bit_array[index] = 1

    def contains(self, item):
        for seed in range(self.hash_functions):
            index = mmh3.hash(item, seed) % self.size
            if self.bit_array[index] == 0:
                return False
        return True
    
with open('mockdata.json', 'r', encoding='utf-8') as file:
    data = json.load(file)
    
bloom_filter = BloomFilter(10000, 4)

for entry in data:
    bloom_filter.add(entry['name'].lower())
    
items_to_check = ['Amit Kumar Sahoo', 'Zelig Joplin', 'John Chataignier', 'Gerik Giurio']
for item in items_to_check:
    if bloom_filter.contains(item.lower()):
        print(f"{item} may exist in the set.")
    else:
        print(f"{item} definitely does not exist in the set.")

File: 004-Bloom-Filter/bloomfilter.js
const fs = require('fs');
const { BloomFilter, PartitionedBloomFilter } = require('bloom-filters');

const rawData = fs.readFileSync('mockdata.json');
const data = JSON.parse(rawData);

const filter = new BloomFilter(10000, 4); // 10000 items, 4 hash functions

data.forEach((entry) => {
    filter.add(entry.name.toLowerCase());
});

const amitExists = filter.has('amit');

if (amitExists) {
    console.log('Amit exists in the names.');
} else {
    console.log('Amit does not exist in the names.');
}

console.log('Error Rate: ', filter.rate())

// Using a partitioned bloom filter
/* A Partitioned Bloom Filter is a variation of a classic Bloom Filter.

This filter works by partitioning the M-sized bit array into k slices of size m = M/k bits, k = nb of hash functions in the filter. Each hash function produces an index over m for its respective slice. Thus, each element is described by exactly k bits, meaning the distribution of false positives is uniform across all elements. */

const partitionedFilter = new PartitionedBloomFilter(10000, 4, 0.1);
// Partitioned Bloom Filter of size 10, 4 hash functions, and a load factor of 0.1

data.forEach((entry) => {
    partitionedFilter.add(entry.name.toLowerCase());
});

const amitExistsInPartitionedFilter = partitionedFilter.has('amit');

if (amitExistsInPartitionedFilter) {
    console.log('Amit exists in the names.(Partitioned)');
} else {
    console.log('Amit does not exist in the names.(Partitioned)');
}

console.log('Error Rate: ', partitionedFilter.rate())